src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs

```diff
@@ -3,106 +3,126 @@
+			{
+				SingleReader = maxConsumers == 1,
+				AllowSynchronousContinuations = true,
+				// wait does not block it simply signals that Writer.TryWrite should return false and be retried
+				// DropWrite will make `TryWrite` always return true, which is not what we want.
+				FullMode = BoundedChannelFullMode.Wait
+			});
+			async Task ConsumeMessages() =>
+				await Consume(options.Throttles.MaxConsumerBufferSize, options.Throttles.MaxConsumerBufferLifeTime).ConfigureAwait(false);
+			for (var i = 0; i < maxConsumers; i++)
+				_backgroundTasks.Add(Task.Factory.StartNew(() => ConsumeMessages(), TaskCreationOptions.LongRunning));
 		}
 
+
 		private IElasticLowLevelClient _lowLevelClient = default!;
 
-		private ElasticsearchLoggerOptions _options = default!;
+		private ElasticsearchLoggerOptions _options;
+		public Channel<LogEvent> PubSub { get; }
+		public ChannelWriter<LogEvent> Writer => PubSub.Writer;
 
-		internal ElasticsearchLoggerOptions Options
+		public void Enqueue(LogEvent item)
 		{
-			get => _options;
-			set
-			{
-				_options = value;
-				UpdateClient();
-			}
+			if (!Writer.TryWrite(item))
+				Options.Throttles.PublishRejectionCallback?.Invoke(item);
 		}
 
-		public void Dispose()
-		{
-			_messageQueue?.CompleteAdding();
-			try
-			{
-				_outputThread.Join(1500); // with timeout in case writer is locked
-			}
-			catch (ThreadStateException) { }
+		private static readonly byte[] LineFeed = { 10 };
 
-			_messageQueue?.Dispose();
-		}
 
-		public void EnqueueMessage(LogEvent logEvent)
+		private async Task Consume(int maxQueuedMessages, TimeSpan maxInterval)
 		{
-			if (!_messageQueue.IsAddingCompleted)
+			using var buffer = new ConsumerBuffer(maxQueuedMessages, maxInterval);
+
+			while (await buffer.WaitToReadAsync(PubSub.Reader).ConfigureAwait(false))
 			{
-				try
+				while (buffer.Count < maxQueuedMessages && PubSub.Reader.TryRead(out LogEvent item))
 				{
-					_messageQueue.Add(logEvent);
-					return;
+					if (buffer.DurationSinceFirstRead > maxInterval) break;
+
+					buffer.Add(item);
 				}
-				catch (InvalidOperationException) { }
-			}
 
-			// Adding is complete, so just log the message
-			try
-			{
-				PostEvent(logEvent);
+				if (buffer.NoThresholdsHit) continue;
+
+				var response = await _lowLevelClient.BulkAsync<DynamicResponse>(
+						PostData.StreamHandler(buffer,
+							(@event, stream) =>
+							{
+								/* NOT USED */
+							},
+							async (@event, stream, ctx) =>
+							{
+								foreach (var logEvent in buffer.Buffer)
+								{
+									var indexTime = logEvent.Timestamp ?? ElasticsearchLoggerProvider.LocalDateTimeProvider();
+									if (_options.IndexOffset.HasValue) indexTime = indexTime.ToOffset(_options.IndexOffset.Value);
+
+									var index = string.Format(_options.Index, indexTime);
+									var indexHeader = new { index = new { _index = index } };
```

Comment:
would there be some benefit to introducing a type for this?

Reply:
There would! Was holding off on dedicated types when following this up with retries, backoff and response failure callbacks.

I'm in two minds to spin off `ElasticsearchDataShipper` as its own package which would benefit from some more typing and control for the end user as well.